src/write.rs

```diff
@@ -124,65 +124,198 @@ fn verify_channel_mask() {
+impl<'w, W: 'w + io::Write + io::Seek> Write for EmbeddedWriter<'w, W> {
+    fn write(&mut self, buf:&[u8]) -> io::Result<usize> {
+        let &mut EmbeddedWriter { ref mut writer, ref mut state } = self;
+        state.write(writer, buf)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        let &mut EmbeddedWriter { ref mut writer, ref mut state } = self;
+        try!(state.update_header(writer));
+        writer.flush()
+    }
+}
+
+impl<'w, W: 'w + io::Write + io::Seek> Drop for EmbeddedWriter<'w, W> {
+    fn drop(&mut self) {
+        let &mut EmbeddedWriter { ref mut writer, ref mut state } = self;
+        let _ignored = state.finalize_chunk(writer);
+    }
+}
+
+#[derive(Copy,Clone)]
+pub struct ChunkWritingState {
+    pub len: u32
+}
+
+impl ChunkWritingState {
+    pub fn write<W: io::Write + io::Seek>(&mut self, writer: &mut W, buf:&[u8]) -> io::Result<usize> {
+        let written = try!(writer.write(buf));
+        self.len += written as u32;
+        Ok(written)
+    }
+
+    pub fn update_header<W: io::Write + io::Seek>(&mut self, writer: &mut W) -> io::Result<()> {
+        try!(writer.seek(io::SeekFrom::End(-(self.len as i64 + 4))));
+        try!(writer.write_le_u32(self.len));
+        try!(writer.seek(io::SeekFrom::End(0)));
+        Ok(())
+    }
+
+    pub fn finalize_chunk<W: io::Write + io::Seek>(&mut self, writer: &mut W) -> io::Result<()> {
+        try!(self.update_header(writer));
+        if self.len % 2 == 1 {
+            try!(writer.write_u8(0));
+        }
+        Ok(())
+    }
+}
+
+/// A Riff chunk Wave writer, allowing to write arbitrary chunks to a file.
+///
+/// For simple out-of-the-box wav usage, prefer the `WavWriter` facade.
+pub struct ChunksWriter<W: io::Write + io::Seek> {
+    /// underlying writer
     writer: W,
+    /// wave spec, if known at that point
+    pub spec_ex: Option<WavSpecEx>,
+    /// state of the data chunk, if currently writing it
+    pub data_state: Option<ChunkWritingState>,
+    dirty: bool,
+    sample_writer_buffer: Vec<u8>,
+}
 
-    /// The number of bytes written to the data section.
+impl<W: io::Write + io::Seek> ChunksWriter<W> {
+    /// Creates a ChunksWriter.
     ///
-    /// This is an `u32` because WAVE cannot accomodate more data.
-    data_bytes_written: u32,
+    /// Write the RIFF header (including a len placeholder). The writer
+    /// is then ready to start writing the first chunk.
+    pub fn new(mut writer: W) -> Result<ChunksWriter<W>> {
+        try!(writer.write_all(b"RIFF\0\0\0\0WAVE"));
+        Ok(ChunksWriter {
+            writer: writer,
+            spec_ex: None,
+            dirty: false,
+            data_state: None,
+            sample_writer_buffer: vec!(),
+        })
+    }
+
+    /// Update the file length field in the RIFF header.
+    ///
+    /// The writer is then repositioned at end of file.
+    fn update_riff_header(&mut self) -> io::Result<()> {
+        let full_len = try!(self.writer.seek(io::SeekFrom::End(0)));
```

Comment:
Also here, we should track the number of bytes written, rather than assuming that the end of the writer coincides with what we wrote.